File: pesquisa_operacional/montecarlosimples.py
import random

def monte_carlo_simples():
    """
    Realiza uma simulação simples de Monte Carlo para estimar a probabilidade
    de um evento com uma dada probabilidade de sucesso.

    A ideia básica é repetir um experimento aleatório muitas vezes e observar
    a frequência com que um determinado resultado ocorre. Essa frequência
    relativa é então usada como uma estimativa da probabilidade do evento.
    """
    num_simulacoes = 10000  # Define o número de vezes que a simulação será executada.
    # Um número maior de simulações tende a fornecer uma estimativa mais precisa
    # da probabilidade real, seguindo a Lei dos Grandes Números.

    eventos_favoraveis = 0  # Inicializa um contador para o número de vezes que o
    # evento de interesse (que definimos como "sucesso") ocorre durante as simulações.

    probabilidade_sucesso = 0.6  # Define a probabilidade teórica de sucesso para cada
    # tentativa individual na simulação. Neste caso, há uma chance de 60% de sucesso.

    print("Simulação de Monte Carlo Simples em Python")
    print(f"Número de Simulações: {num_simulacoes}")
    print(f"Probabilidade de Sucesso em cada tentativa: {probabilidade_sucesso}")
    print("-" * 30)

    # Inicia o loop principal da simulação. Cada iteração representa uma tentativa
    # independente do evento que estamos modelando.
    for _ in range(num_simulacoes):
        # Gera um número de ponto flutuante aleatório no intervalo [0.0, 1.0).
        # Este número aleatório simula o resultado de uma única tentativa.
        resultado_aleatorio = random.random()

        # Verifica se o resultado aleatório é menor que a probabilidade de sucesso.
        # Se for, consideramos que a tentativa foi um "evento favorável" (sucesso).
        if resultado_aleatorio < probabilidade_sucesso:
            eventos_favoraveis += 1  # Incrementa o contador de eventos favoráveis.

    # Após a conclusão de todas as simulações, calculamos a probabilidade estimada
    # do evento favorável ocorrer. Esta estimativa é a proporção do número de
    # eventos favoráveis observados em relação ao número total de simulações realizadas.
    probabilidade_estimada = eventos_favoraveis / num_simulacoes

    # Imprime os resultados finais da simulação para o usuário.
    # Isso inclui o número total de eventos favoráveis que foram contados
    # e a probabilidade estimada do evento com base na simulação de Monte Carlo.
    print(f"Número de Eventos Favoráveis: {eventos_favoraveis}")
    print(f"Probabilidade Estimada do Evento Favorável: {probabilidade_estimada}")

# Este bloco de código garante que a função 'monte_carlo_simples()' seja chamada
# apenas quando o script Python é executado diretamente (e não quando é importado
# como um módulo em outro script).
if __name__ == "__main__":
    monte_carlo_simples()

File: pesquisa_operacional/MonteCarloSimples.java
import java.util.Random;

public class MonteCarloSimples {

    public static void main(String[] args) {
        // Define o número de simulações que serão realizadas.
        // Um número maior de simulações geralmente leva a uma estimativa mais precisa,
        // mas também aumenta o tempo de execução do programa.
        int numSimulacoes = 10000;

        // Cria uma instância da classe Random para gerar números aleatórios.
        // Esta classe fornece métodos para gerar diferentes tipos de números aleatórios.
        Random random = new Random();

        // Inicializa um contador para rastrear o número de vezes que um "evento favorável" ocorre.
        // A definição de "evento favorável" é específica para o problema que estamos simulando.
        int eventosFavoraveis = 0;

        // Define um parâmetro crucial para a nossa simulação.
        // Neste exemplo, estamos simulando um evento com uma probabilidade de ocorrer
        // de 60% (ou 0.6) em cada tentativa independente.
        double probabilidadeSucesso = 0.6;

        // Imprime informações iniciais sobre a simulação para o usuário.
        System.out.println("Simulação de Monte Carlo Simples");
        System.out.println("Número de Simulações: " + numSimulacoes);
        System.out.println("Probabilidade de Sucesso em cada tentativa: " + probabilidadeSucesso);
        System.out.println("----------------------------------");

        // Este loop for executa o número de simulações definido.
        // Cada iteração representa uma tentativa ou uma execução do processo que estamos modelando.
        for (int i = 0; i < numSimulacoes; i++) {
            // Dentro de cada simulação, geramos um número aleatório entre 0.0 (inclusive) e 1.0 (exclusive).
            // Este número aleatório é usado para determinar se um "evento favorável" ocorreu nesta tentativa.
            if (random.nextDouble() < probabilidadeSucesso) {
                // Se o número aleatório gerado for menor que a probabilidade de sucesso definida,
                // consideramos que um "evento favorável" ocorreu e incrementamos o contador.
                eventosFavoraveis++;
            }
        }

        // Após a conclusão de todas as simulações, calculamos a probabilidade estimada
        // do "evento favorável" ocorrer. Fazemos isso dividindo o número total de
        // eventos favoráveis observados pelo número total de simulações realizadas.
        double probabilidadeEstimada = (double) eventosFavoraveis / numSimulacoes;

        // Imprime os resultados finais da simulação para o usuário.
        // Isso inclui o número total de eventos favoráveis encontrados e a probabilidade estimada.
        System.out.println("Número de Eventos Favoráveis: " + eventosFavoraveis);
        System.out.println("Probabilidade Estimada do Evento Favorável: " + probabilidadeEstimada);
    }
}
